transport/transport.ts

```
@@ -515,3 +476,220 @@ export abstract class Transport<ConnType extends Connection> {
+          } -- received invalid handshake resp: ${JSON.stringify(parsed)}`,
+        );
+        return;
+      }
+
+      if (!parsed.payload.status.ok) {
+        log?.warn(
+          `${this.clientId} -- received failed handshake resp: ${JSON.stringify(
+            parsed,
+          )}`,
+        );
+        return;
+      }
+
+      // everything is ok
+      // connect the session
+      const session = this.onConnect(conn, parsed.from);
+      sessionCb?.(session);
+      conn.removeDataListener(bootHandler);
+      conn.addDataListener((data) => this.handleMsg(this.parseMsg(data)));
+    };
+
+    return bootHandler;
+  }
+
+  onDisconnect(conn: ConnType, connectedTo: string | undefined): void {
+    if (connectedTo) this.inflightConnectionPromises.delete(connectedTo);
+    super.onDisconnect(conn, connectedTo);
+  }
+}
+
+export abstract class ServerTransport<
+  ConnType extends Connection,
+> extends Transport<ConnType> {
+  protected handleConnection(conn: ConnType) {
+    let session: Session<ConnType> | undefined = undefined;
+    const client = () => session?.connectedTo ?? 'unknown';
+    conn.addDataListener(
+      this.receiveWithBootSequence(conn, (establishedSession) => {
+        session = establishedSession;
+      }),
+    );
+
+    conn.addCloseListener(() => {
+      if (!session) return;
+      log?.info(
+        `${this.clientId} -- connection (id: ${
+          conn.debugId
+        }) to ${client()} disconnected`,
+      );
+      this.onDisconnect(conn, session?.connectedTo);
+    });
+
+    conn.addErrorListener((err) => {
+      if (!session) return;
+      log?.warn(
+        `${this.clientId} -- connection (id: ${
+          conn.debugId
+        }) to ${client()} got an error: ${err}`,
+      );
+    });
+  }
+
+  receiveWithBootSequence(
+    conn: ConnType,
+    sessionCb?: (sess: Session<ConnType>) => void,
+  ) {
+    const bootHandler = (data: Uint8Array) => {
+      const parsed = this.parseMsg(data);
```

I'm still very uncomfortable that failing to read/parse a message doesn't just break everything and silently fails.

---

hm do you think this should be fatal?

---

Yeah probably, I mean if we get a bad packet, we'll probably continue getting bad ones, right? Like you can't recover that packet and we'll likely continue to receive bad packets

---

gotcha, i'll do that in a later PR after we merge this stack